[DOCS][ESQL][8.14] Add API key based security model info for ESQL CCS (#109155)
elastic/elasticsearch: docs/reference/esql/esql-across-clusters.asciidoc

include::{es-ref-dir}/search/search-your-data/search-across-clusters.asciidoc[tag=ccs-proxy-mode]

[discrete]
[[esql-ccs-security-model]]
==== Security model

[discrete]
[[esql-ccs-security-model-api-key]]

> ===== API key authentication

Reviewer: Can you also add a section for "TLS certificate authentication". Probably just a small paragraph that links to https://www.elastic.co/guide/en/elasticsearch/reference/current/remote-clusters-cert.html and mentions that same named role must exist on in both clusters and it is recommended that the same named roles across clusters have identical privileges.  Don't want to restate too much here, mostly just calling out that 2 models exist.

Author: Makes sense 👍

[NOTE]
====
The `ENRICH` keyword is *not supported* until 8.15 when using {esql} with the API key based security model.
====

The following information pertains to using {esql} across clusters with the <<remote-clusters-api-key, *API key based security model*>>.
API key based cross-cluster search (CCS) enables more granular control over allowed actions between clusters. 

You will need to create an API key using the <<security-api-create-cross-cluster-api-key,Create cross-cluster API key>> API or using the {kibana-ref}/api-keys.html[Kibana API keys UI].

The following example API call creates a role that can query remote indices using {esql} with the API key based security model.

[source,console]
----
POST /_security/role/remote1
{
  "indices": [
    { 
      "names" : [""], <2>
      "privileges": ["read"]
    }
  ], 
  "remote_indices": [
    {

>       "names": [ "logs-*" ],

Reviewer: can add a call out here. it's kinda self evident, but I think spelling it out can help. 
for example, 
```
<3> The indices allowed be read in the remote cluster. The configured <<?? how to configure the cross cluster api key ??, cross-cluster API key>> must also allow this index to be read. 
```

      "privileges": [ "read","read_cross_cluster" ], <1>

>       "clusters" : ["my_remote_cluster"]

Reviewer: can add a call out here too ? i.e.
```
<4> The remote clusters to which these privileges apply. This remote cluster must be configured <<?how to configure the cross cluster api key??, cross-cluster API key>>  and connected << link to remote info api>> before the remote index can be queried.

include::{es-ref-dir}/search/search-your-data/search-across-clusters.asciidoc[tag=ccs-proxy-mode]

[discrete]
[[esql-ccs-security-model]]
==== Security model

{es} supports two security models for cross-cluster search (CCS):

* <<esql-ccs-security-model-certificate, TLS certificate authentication>>
* <<esql-ccs-security-model-api-key, API key authentication>>

[discrete]
[[esql-ccs-security-model-certificate]]
===== TLS certificate authentication

> TLS certificate authentication makes sense in single administrator scenarios where you have full control over both clusters.

Author: Do we talk about licenses explicitly in docs given variance between self-managed and cloud?

> TLS certificate authentication makes sense in single administrator scenarios where you have full control over both clusters.

Reviewer: I don't know our general stance, but it seems like an important detail since ES|QL is not gated by an enterprise license. IIUC in cloud user have access to all features, but are billed based on usage.

> TLS certificate authentication makes sense in single administrator scenarios where you have full control over both clusters.

Author: We use generic messaging about licensing as a rule, and the main pages for these security models have prerequisites sections that say "_The local and remote clusters must have an appropriate license. For more information, refer to https://www.elastic.co/subscriptions_". So I will refrain from mentioning licensing in this doc.

> Importantly, you must ensure that named roles and their privileges are identical in both clusters.

Author: What's the reason for not keeping strong recommendation that they be strictly identical?

> Importantly, you must ensure that named roles and their privileges are identical in both clusters.

Reviewer: "must ensure...identical" is not technically accurate.  recommending that they exist and are identical is sane approach, but it is not technically required. e.g. if you don't it will still work, but the results may be surprising. there are likely some cases where you intentionally omit some permissions across clusters.

Refer to <<remote-clusters-cert, TLS certificate authentication>> for prerequisites and detailed setup instructions.

[discrete]
[[esql-ccs-security-model-api-key]]
===== API key authentication

[NOTE]
====
`ENRICH` is *not supported* in this version when using {esql} with the API key based security model.
====

The following information pertains to using {esql} across clusters with the <<remote-clusters-api-key, *API key based security model*>>.
API key based cross-cluster search (CCS) enables more granular control over allowed actions between clusters.

> As a result, defining roles and privileges is more complex than with the certificate based security model.

Author: Can basically remove this sentence and lean on previous one, if we don't want my non-specific signpost that API route is more complex :)

> As a result, defining roles and privileges is more complex than with the certificate based security model.

Reviewer: Just trying to avoid calling one more complex than the other. Keeping roles in sync is also complex, just in a different way. Users may want to adopt the API key based model due to the complexities of mirroring roles (we don't provide any tooling to do that).

For example, it makes sense in scenarios when you have different administrators for different clusters.

> You will need to create an API key using the <<security-api-create-cross-cluster-api-key,Create cross-cluster API key>> API or using the {kibana-ref}/api-keys.html[Kibana API keys UI].

Reviewer: ```suggestion
For the cluster that has the data (the remote cluster), you will need to create an API key using the <<security-api-create-cross-cluster-api-key,Create cross-cluster API key>> API or using the {kibana-ref}/api-keys.html[Kibana API keys UI] and then <<remote-clusters-api-key, add the remote cluster>>.
```

> The following example API call creates a role that can query remote indices using {esql} with the API key based security model.

Reviewer: ```suggestion
Using {esql} with the API key based security model requires some additional permissions that may not be needed with query dsl based search. The following example API call creates a role that can query remote indices using {esql} with the API key based security model.
```

[source,console]
----
POST /_security/role/remote1
{

>   "cluster": ["cross_cluster_search"], <2>

Reviewer: nit: can these be re-ordered to read top down 1-5 ?

Author: totally was trying to give a sense of relative importance but agree might be a net bad reading experience

Refer to <<remote-clusters-cert, TLS certificate authentication>> for prerequisites and detailed setup instructions.

[discrete]
[[esql-ccs-security-model-api-key]]
===== API key authentication

[NOTE]
====
`ENRICH` is *not supported* in this version when using {esql} with the API key based security model.
====

The following information pertains to using {esql} across clusters with the <<remote-clusters-api-key, *API key based security model*>>. You'll need to follow the steps on that page for the *full setup instructions*. This page only contains additional information specific to {esql}.

API key based cross-cluster search (CCS) enables more granular control over allowed actions between clusters.
This may be the preferred model when you have different administrators for different clusters and want more control over who can access what data. In this model, cluster administrators must explicitly define the access given to clusters and users.

You will need to:

* Create an API key on the *remote cluster* using the <<security-api-create-cross-cluster-api-key,Create cross-cluster API key>> API or using the {kibana-ref}/api-keys.html[Kibana API keys UI].

> * Add the API key to the `elasticsearch.yml` file on the *local cluster*, as part of the steps in <<remote-clusters-security-api-key-local-actions,configuring the local cluster>>. All cross-cluster requests from the local cluster are bound by the API key’s privileges.

Reviewer: ```suggestion
* Add the API key to the keystore on the *local cluster*, as part of the steps in <<remote-clusters-security-api-key-local-actions,configuring the local cluster>>. All cross-cluster requests from the local cluster are bound by the API key’s privileges.
```

----
POST /_security/role/remote1
{
  "cluster": ["cross_cluster_search"], <2>
  "indices": [
    { 
      "names" : [""], <3>
      "privileges": ["read"]
    }
  ], 
  "remote_indices": [ <4>
    {
      "names": [ "logs-*" ],
      "privileges": [ "read","read_cross_cluster" ], <1>
      "clusters" : ["my_remote_cluster"] <5>
    }
  ]
}
----

<1> The `read_cross_cluster` privilege is always required when using {esql} across clusters with the API key based security model.

> <2> The `cross_cluster_search` cluster privilege is also required.

Author: ```suggestion
<2> The `cross_cluster_search` cluster privilege is also required for the _local_ cluster.
```
